### PR TITLE
Dialog

### LINKED ISSUE #481: Autofocus on MBDialog causes an error when the focussed element is a MBNumericDoubleField
When the first lement inside the MBDialog body is a MBNumericDoubleField, the auto-focus seems to cause an exception in the JS.
When I put any other focusable element before the MBNumericDoubleField, it does not throw the exception.
Capture2

### PR DESCRIPTION
**Pull request by stefanloerwald**
This is a bigger change for how dialogs are handled. I'll add comments to explain the changes.

### CODE REVIEW

File: Material.Blazor.Website/Pages/Dialog.razor [hunk 1]
```diff
@@ -203,41 +203,44 @@
...
-
-                    <div class="mdc-layout-grid__inner">
-                        <div class="mdc-layout-grid__cell--span-12">
-                            <MBTextField @bind-Value="@UserLogon.UserID"
-                                         Label="User ID"
-                                         style="width: 100%; margin-top: 6px;"
-                                         TextInputStyle="@MBTextInputStyle.Outlined"
-                                         TrailingIcon="person"
-                                         ValidationMessageFor="@(() => UserLogon.UserID)" />
+                @if (UserLogon != null)
```
Comment (stefanloerwald) [inline Material.Blazor.Website/Pages/Dialog.razor]: This is now necessary, because dialogs are added to the DOM sooner.

TODO: make a remark in the Changelog, because this might affect some users.

File: Material.Blazor/Components/DatePicker/InternalDatePickerPanel.razor.cs [hunk 2]
```diff
@@ -213,6 +213,7 @@ private void OnNextMonthClick()
         /// <inheritdoc/>
         protected override async Task OnAfterRenderAsync(bool firstRender) // TODO: this prevents us from marking OnAfterRenderAsync in InputComponent as sealed. Consider alternatives!
         {
+            await base.OnAfterRenderAsync(firstRender);
```
Comment (stefanloerwald) [inline Material.Blazor/Components/DatePicker/InternalDatePickerPanel.razor.cs:216]: We perform relevant work in base.OnAfterRenderAsync, so we can't simply omit that.

File: Material.Blazor/Components/Dialog/MBDialog.razor [hunk 3]
```diff
@@ -1,49 +1,44 @@
...
 @inherits ComponentFoundation
 
+<div @ref="DialogElem"
+     class="mdc-dialog @ActiveConditionalClasses @(@class)"
+     style="@style"
+     id="@id"
+     @attributes="@MyAttributes"
+     aria-modal="true">
 
-@if (IsOpen)
```
Comment (stefanloerwald) [inline Material.Blazor/Components/Dialog/MBDialog.razor:5]: Most importantly, the dialog is now always rendered. This causes the indentation difference, which makes this look like a huge change

File: Material.Blazor/Components/Dialog/MBDialog.razor [hunk 4]
```diff
@@ -1,49 +1,44 @@
...
-{
-    <div @key="@Key"
```
Comment (stefanloerwald) [inline Material.Blazor/Components/Dialog/MBDialog.razor:7]: This is the only actual other change: no more key. When we re-open the same dialog, the child elements were already initialized correctly once, so no need to discard that entire dom tree.

File: Material.Blazor/Components/Dialog/MBDialog.razor.cs [hunk 5]
```diff
@@ -1,20 +1,16 @@
...
 using System.Threading.Tasks;
 
 namespace Material.Blazor
 {
     /// <summary>
-    /// This is a general purpose Material Theme dialog. The Material.Blazor dialog implementation does not render a dialog at all until
-    /// it is shown by the consumer. At this stage it is rendered and Material Theme is permitted to control it's opening and to 
-    /// intiate it and its embedded components. Likewise when the dialog is closed, Material Theme is permitted to close the dialog
-    /// gracefully before Material.Blazor removes the markup.
+    /// This is a general purpose Material Theme dialog.
```
Comment (stefanloerwald) [inline Material.Blazor/Components/Dialog/MBDialog.razor.cs:9]: We don't do that late-adding anymore, so comment adjusted

File: Material.Blazor/Components/Dialog/MBDialog.razor.cs [hunk 6]
```diff
@@ -65,18 +61,11 @@ public partial class MBDialog : ComponentFoundation, IMBDialog
         [Parameter] public bool OverflowVisible { get; set; } = false;
 
 
-        /// <summary>
```
Comment (stefanloerwald) [inline Material.Blazor/Components/Dialog/MBDialog.razor.cs:68]: Dialog instantiation is now flagged with a TaskCompletionSource

File: Material.Blazor/Components/Dialog/MBDialog.razor.cs [hunk 7]
```diff
@@ -65,18 +61,11 @@ public partial class MBDialog : ComponentFoundation, IMBDialog
...
-        /// </summary>
-        bool IMBDialog.DialogHasInstantiated => dialogHasInstantiated;
-
-
         private ElementReference DialogElem { get; set; }
         private bool HasBody => Body != null;
         private bool HasButtons => Buttons != null;
         private bool HasCustomHeader => CustomHeader != null;
         private bool HasTitle => !string.IsNullOrWhiteSpace(Title);
-        private List<DialogChildComponent> LayoutChildren { get; set; } = new List<DialogChildComponent>();
```
Comment (stefanloerwald) [inline Material.Blazor/Components/Dialog/MBDialog.razor.cs:79]: The child components themselves wait for the dialog to be opened, so no need to list them here

File: Material.Blazor/Components/Dialog/MBDialog.razor.cs [hunk 8]
```diff
@@ -182,73 +153,25 @@ public async Task<string> ShowAsync()
...
+        private readonly TaskCompletionSource OpenedSource = new();
+        internal Task Opened => OpenedSource.Task;
+
 
         [JSInvokable]
         public void NotifyOpened()
         {
-            AfterDialogInitialization = true;
-            StateHasChanged();
+            OpenedSource.TrySetResult();
```
Comment (stefanloerwald) [inline Material.Blazor/Components/Dialog/MBDialog.razor.cs]: This essentially "releases" a semaphore (without being a semaphore). Components within the dialog will perform no JS interop until this is set.

File: Material.Blazor/Components/IconButtonToggle/MBIconButtonToggle.razor.cs [hunk 9]
```diff
@@ -86,7 +86,7 @@ private void ToggleOnClick()
         /// </summary>
         /// <param name="sender"></param>
         /// <param name="e"></param>
-        protected void OnDisabledSetCallback() => InvokeAsync(() => AllowNextRender = true);
+        protected void OnDisabledSetCallback() => InvokeAsync(AllowNextShouldRender);
```
Comment (stefanloerwald) [inline Material.Blazor/Components/IconButtonToggle/MBIconButtonToggle.razor.cs:89]: minor unrelated change: use the method instead of field.

File: Material.Blazor/Foundation/ComponentFoundation.cs [hunk 10]
```diff
@@ -23,7 +23,9 @@ public abstract class ComponentFoundation : ComponentBase, IDisposable
         private static readonly ImmutableArray<string> EssentialSplattableAttributes = ImmutableArray.Create("formnovalidate", "max", "min", "role", "step", "tabindex", "type", "data-prev-page");
         private bool? disabled = null;
 
-        [Inject] private protected IBatchingJSRuntime JsRuntime { get; set; }
+        [Inject] private IBatchingJSRuntime InjectedJsRuntime { get; set; }
+        protected IBatchingJSRuntime JsRuntime { get; set; }
+        [CascadingParameter] private MBDialog ParentDialog { get; set; }
```
Comment (stefanloerwald) [inline Material.Blazor/Foundation/ComponentFoundation.cs:28]: If we're inside a dialog, we'll know with this cascading parameter. We "override" the JsRuntime in that case

File: Material.Blazor/Foundation/ComponentFoundation.cs [hunk 11]
```diff
@@ -222,6 +224,7 @@ protected sealed override void OnParametersSet()
         /// </summary>
         protected sealed override void OnInitialized()
         {
+            JsRuntime = ParentDialog == null ? InjectedJsRuntime : new DialogAwareBatchingJSRuntime(InjectedJsRuntime, ParentDialog);
```
Comment (stefanloerwald) [inline Material.Blazor/Foundation/ComponentFoundation.cs:227]: If inside a dialog, we create a wrapper around the JsRuntime

File: Material.Blazor/Foundation/DialogAwareBatchingJSRuntime.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        private readonly IBatchingJSRuntime underlyingJSRuntime;
+        private readonly MBDialog dialog;
+        public DialogAwareBatchingJSRuntime(IBatchingJSRuntime underlyingJSRuntime, MBDialog dialog)
+        {
+            this.underlyingJSRuntime = underlyingJSRuntime;
+            this.dialog = dialog;
+        }
+        public async Task<T> InvokeAsync<T>(string identifier, params object[] args)
+        {
+            await dialog.Opened;
```
Comment (stefanloerwald) [inline Material.Blazor/Foundation/DialogAwareBatchingJSRuntime.cs:16]: For a JS call on a component that is within a dialog, this code now handles the JS interop, so we will always wait until the dialog is actually open.

File: Material.Blazor/Foundation/InputComponent.cs [hunk 13]
```diff
@@ -183,7 +176,7 @@ protected string ComponentValueAsString
         /// <summary>
         /// Allows <see cref="ShouldRender()"/> to return "true" for the next render only.
         /// </summary>
-        internal bool AllowNextRender = false;
+        private bool AllowNextRender = false;
```
Comment (stefanloerwald) [inline Material.Blazor/Foundation/InputComponent.cs:179]: minor unrelated change: use the method instead of field.

File: Material.Blazor/Foundation/SingleSelectComponent.cs [hunk 14]
```diff
@@ -41,8 +41,7 @@ public abstract class SingleSelectComponent<T, TListElement> : InputComponent<T>
                         }
                     }
 
-                    AllowNextRender = true;
-                    InstantiateAfterNextRender = true;
+                    AllowNextShouldRender();
```
Comment (stefanloerwald) [inline Material.Blazor/Foundation/SingleSelectComponent.cs:44]: minor unrelated change: use the method instead of field.

File: Material.Blazor/Foundation/SingleSelectComponent.cs [hunk 15]
```diff
@@ -93,7 +92,7 @@ public abstract class SingleSelectComponent<T, TListElement> : InputComponent<T>
                     case MBItemValidation.DefaultToFirst:
                         var defaultValue = items.FirstOrDefault().SelectedValue;
                         _ = ValueChanged.InvokeAsync(defaultValue);
-                        AllowNextRender = true;
+                        AllowNextShouldRender();
```
Comment (stefanloerwald) [inline Material.Blazor/Foundation/SingleSelectComponent.cs:95]: minor unrelated change: use the method instead of field.

File: Material.Blazor/Foundation/InputComponent.cs [hunk 16]
```diff
@@ -347,22 +316,6 @@ protected sealed override bool ShouldRender()
         }
 
 
-        /// <summary>
-        /// Material.Blazor components descending from <see cref="InputComponent{T}"/> _*must not*_ override OnAfterRenderAsync(bool).
-        /// </summary>
-        protected override async Task OnAfterRenderAsync(bool firstRender)
```
Comment (stefanloerwald) [inline Material.Blazor/Foundation/InputComponent.cs:353]: Obsolete now: no more need for delaying the initialization. This caused problems anyway, because this was the _only_ interop call we delayed, all other ones weren't.

**Comment by stefanloerwald** [comment]
There's a regression in checkbox, since a parameter setter invokes JS, which happens before the component is correctly initialized. I'll fix that.

File: Material.Blazor.Website/Pages/Dialog.razor (hunk 1 above)
Comment (stefanloerwald) [reply Material.Blazor.Website/Pages/Dialog.razor]: TODO: test whether this change is really still necessary

File: Material.Blazor/Components/Dialog/MBDialog.razor [hunk 17]
```diff
@@ -1,19 +1,19 @@
...
-         style="@style"
-         id="@id"
-         @attributes="@MyAttributes"
-         aria-modal="true">
-
-        <CascadingValue Value="@this" IsFixed="true">
-            <div class="mdc-dialog__container">
+    <CascadingValue Value="@this" IsFixed="true">
+        <div class="mdc-dialog__container">
+            @if (IsOpen)
```
Comment (stefanloerwald) [inline Material.Blazor/Components/Dialog/MBDialog.razor:15]: Only the dialog content is now not added to the DOM, until the user actually want's this dialog to be shown

**Review by MarkStega — APPROVED** [review]
I've tested all dialogs in Optimi$er and all appear ti=o be working as they did before these changes.
